portfolio/src/main/webapp/script.js

```diff
@@ -109,22 +178,164 @@ function retrieveId(string, source){
 }
 
 /**
-  * Runs the search by getting the source HTML, getting the appropriate tag, and scrolling towards that id
-  */
+ * Get's all id's from a given tag array
+ */
+function getIds(array, source){
+    var retArr = [];
+    for (var i = 0; i < array.length; i++){
+        var retriever = retrieveId(array[i], source);
+        if (retriever === "nothing"){
+            continue;
+        } else {
+            retArr.push(retriever);
+        }
+    }
+
+    return retArr;
+}
+
+/**
+ * Counts the number of duplicate elements in an array
+ * NOT USED
+ */
+function getCounts(array){
+
+    var counts = {};
+    array.forEach(function(x) { 
+        counts[x] = (counts[x] || 0) + 1; 
+    });
+    
+    return counts;
+}
+
+/**
+ * Wraps each matching element in a div in a span class
+ */
+function wrapInSpan(div, wantedWord) {
+
+	var div = document.getElementById(div);
+    var text = div.innerHTML.toLocaleLowerCase();
+
+    //split spaces, commas, quotes
+    var wordArr = text.split(/[\s,"]+/);
+
+    text = [];
+    wordArr.forEach(function(el) {
+        if(el.includes(wantedWord)){    
+            el = '<span class="highlight" id="spanned">' + el + '</span>';
+        }
+        text.push(el);
+    });
+
+    text = text.join(' ');
+    div.innerHTML = text;
+}
+
+/**
+ *  Reverts given div back to original source code
+ */
+function clearDivAtIndex(div){
+    document.getElementById(div).innerHTML = documentCopy.getElementById(div).innerHTML; 
+}
+
+/**
+ *  Removes duplicates in an array
+ */
+function removeDuplicates(array){
+    newArr = [...new Set(array)];
+    return newArr;
+}
+
+/**
+ * Handles physically highlighting each match and clearing matches prior
+ */
+function runHighlights(ids, highlightedWord, src, index){
+    newids = removeDuplicates(ids);
+
+    //if done with all matches
+    if (index === newids.length){
+        clearDivAtIndex(newids[index-1]);
+        setHidden('nextbutton', true);
+        document.body.scrollIntoView({behavior: "smooth"}); //back to the top
+        return;
+    }
+    
+    //clear the prior div
+    if (index > 0){
+        clearDivAtIndex(newids[index-1]);
+    }
+
+    wrapInSpan(newids[index], highlightedWord);
+
+    // no support for safari or ie
+    document.getElementById(newids[index]).scrollIntoView({behavior: "smooth"}); 
+}
+
 function runSearch(input){
+    
```

Comment:
nit: remove lines at the beginning and end of the function. same for other functions

some guidance on that: https://google.github.io/styleguide/jsguide.html#formatting-vertical-whitespace